ENH: Add `scipy.datasets` submodule

AnirudhDagar opened the pull request:
#### Reference PR
gh-8707

#### What does this implement/fix?

```
>>> from scipy import datasets
# Example ascent dataset loading with the new module
>>> datasets.ascent()
array([[ 83,  83,  83, ..., 117, 117, 117],
       [ 82,  82,  83, ..., 117, 117, 117],
       [ 80,  81,  83, ..., 117, 117, 117],
       ...,
       [178, 178, 178, ...,  57,  59,  57],
       [178, 178, 178, ...,  56,  57,  57],
       [178, 178, 178, ...,  57,  57,  58]])
```

With gh-8707, a couple of years ago, SciPy wanted to introduce the `datasets` submodule and move a handful of dataset functions from the current `misc` module to this new `datasets` submodule. Thanks to @WarrenWeckesser for the initial work. With this PR (indeed inspired by gh-8707) we resume those efforts after making some improvements (explained below) and move away from the `scipy.misc` module, finally deprecating it (in a separate PR #15901).

- [x] Add `scipy.datasets` submodule
- [x] Utilize [pooch](https://github.com/fatiando/pooch) to handle the dataset downloading and caching.
- [x] Enable meson support for `scipy.datasets` submodule
- [x] Update the base URL and registry URL dataset links once we decide the approach and create the respected repos in the scipy organization.

TODO Later:
- [ ] Move all dataset files (eg. `scipy.stats` has its own [test datasets](https://github.com/scipy/scipy/tree/main/scipy/stats/tests/data) within the repo) to their respective new repos (explained below). This is something that can be done after landing this PR once we have a concrete datasets API and approach defined for adding new datasets.
- [x] Deprecate the misc module (#15901)

#### Additional information

Why should we use Pooch?
OR
What are some other possible options (eg. intake, one of the suggestions made earlier) and why not prefer them?

There are a few good reasons for choosing Pooch:

1. With Pooch we can easily decouple the datasets that are currently present within the SicPy repository and move them to their own new repository. For example, see https://github.com/scipy-datasets, where each dataset has its own repository. This will lead to a lightweight SciPy wheel decreasing the download size for future releases. Keeping the datasets in individual repos or a single `scipy-datasets` repo is a point of discussion. Note that we need to take into account that GitHub has a 2Gb limit on the repo size. In case, in future, we plan to add new datasets which are relatively large, then this idea of keeping the datasets in the SciPy repo may not be a wise choice. In fact, keeping all the datasets in a separate single dataset repo might also break after a while. The best approach IMO would be to maybe group them by type of datasets into a few repositories or just keep them individually (i.e each dataset has its own repo).

2. Dependency: Pooch is an extremely light package and has only a few [dependencies](https://github.com/fatiando/pooch/blob/main/setup.cfg#L43-L46), so if we were to add a new dependency i.e. Pooch, we can expect it to be small and at the same time it won't add a lot of sub-dependencies. On the other hand, something like intake has a few more deps, see [requirements.txt](https://github.com/intake/intake/blob/master/requirements.txt). With the basic intake package and some additional sections, it will also install packages like `dask` etc.(not required for scipy datasets). Both of them use a library called `appdirs` which is quite small and useful.

3. IMO Pooch has a simpler API overall and seems easy to implement/maintain for developers while also supporting the key features like caching (hash checks, multi download). Intake is slightly more complex with its design of catalog like YAML files and definitely has some extra features but we surely don't need those to solve the problem of dataset downloading and caching.

Ps. [`scikit-image`](https://github.com/scikit-image/scikit-image/) also recently (in 2020) made a move to make use of Pooch for some of the same reasons.

cc @rgommers

stefanv commented:
/cc @grlee77 @hmaarrfk Any tips from [the skimage fetcher](https://github.com/scikit-image/scikit-image/blob/main/skimage/data/_fetchers.py) that should apply here?

rgommers commented:
> 2\. Dependency: Pooch is an extremely light package and has only a few [dependencies](https://github.com/fatiando/pooch/blob/main/setup.cfg#L43-L46), so if we were to add a new dependency i.e. Pooch, we can expect it to be small and at the same time it won't add a lot of sub-dependencies

Scikit-image doesn't depend on Pooch, it's an optional dependency (e.g., `pip install scikit-image[data]`). We should probably do something similar, dragging it in by default seems like a bit much. xref https://github.com/scikit-image/scikit-image/pull/5105 for the relevant discussion.

ilayn commented:
Caching and downloading parts can be done without the dependencies. Requests is quite a hefty dependence for a light weight lib in my opinion. We can simply check a file with a predefined hash and otherwise get it with some urllib machinary.

Also if scikit image is providing these then probably we should delegate to it since they are the golden standard for image related stuff.

Note however that we need a long deprecation time as we witnessed before with imresize and other related functions that haunted us for at least four major versions.

andyfaff commented:
I'm with Ralf on the dependencies, I definitely wouldn't want to be adding more hard dependencies, only optional ones (if at all).

In any case, it's probably worth asking ourselves "Would we add the same files to the project nowadays?" and "Do we think it's worthwhile to add more datasets/images?". If the answer to these questions is no then why don't we just deprecate and remove them completely?

ev-br commented:
What Andrew said. If we go down to `pip install scipy[data]` or some such, why not go all the way to a separate `scipy_datasets` repo/project in the scipy organization.

rgommers commented:
> In any case, it's probably worth asking ourselves "Would we add the same files to the project nowadays?" and "Do we think it's worthwhile to add more datasets/images?". If the answer to these questions is no then why don't we just deprecate and remove them completely?

That's a very fair question. I think the answer is that we do actually want these for examples, tutorials and tests, and ideally want some more. 

For image data we could say why not just use `skimage.data` (https://scikit-image.org/docs/stable/api/skimage.data.html), and add them there? This has a few pros and cons:
- Pro: less to maintain for us
- Con: need to depend on scikit-image in docs and test suites, which is a heavier dependency, circular, and drags in 6 more packages (https://github.com/scikit-image/scikit-image/blob/main/requirements/default.txt) in addition to still depending on `pooch` anyway

In addtion, we also want:
-  time series data (like `electrocardiagram`)
- stats data (like `iris`, and test/benchmark data like the NIST data we already ship)
- `optimize` benchmark data, like the Netlib LP data in `benchmarks/benchmarks/linprog_benchmark_files/`
- all the special functions reference data in `scipy/special/tests/data/`

@AnirudhDagar didn't really focus on the yes/no for `scipy.datasets` here, because the linked PR from @WarrenWeckesser got thorough and largely positive feedback (https://github.com/scipy/scipy/pull/8707#issuecomment-380372344) which still applies. In addition, I think the importance of sdist/wheel size reduction has grown since then, and there's more to gain than was pointed out in that thread. I recently counted, and came to about 18 MB of text/data files that we ship (uncompressed, didn't carefully measure how much wheel sizes reduce with all that removed).

So my impression is that:
- the case for having a single place for all our datasets needs is strong,
- depending on scikit-image for image data, scikit-learn for iris, etc. is not a good solution,
- so the choice is between `scipy.datasets` and a separate `scipy-datasets`. for that:
  - functionality is the same either way
  - we need it to make `scipy.test()` work
  - it's more discoverable as part of the main scipy docs
  - release/docs/maintenance-wise it's easier to have it integrated rather than having an optional dependency on a (pinned version of) `scipy-datasets`.

Regarding file sizes: my impression is that >95% of users will not need `scipy.datasets` contents, unless it becomes much more popular in the future, so we can assume this is disk space & network bandwidth saved. I did a little calculation for PyPI:
- assuming 18 MB on disk may translate to ~3 MB wheel size reduction, then with the current 35M downloads/months we save `36e6 * 12 * 3e6 / 1e15  # downloads/month * months * wheel-MB / PB = 1.3 PB/yr`
- PyPI bandwidth usage is about 1 PB/day (see https://dustingram.com/articles/2021/04/14/powering-the-python-package-index-in-2021/)
- So it's about 0.35% of total PyPI bandwidth size (a little handwaving, but I'm reasonably sure it's >0.1% unless I made a calculation error)

charris commented:
I expect the sdists need to include tests, but do the binary wheels need them?

rgommers commented:
> I expect the sdists need to include tests, but do the binary wheels need them?

We've always wanted this so users can actually diagnose install issues, and I personally use it a lot as well. It could be separated out as another package (e.g., `scipy-testsuite`), but also that is quite a bit of work. The main downside of actually including tests is data size, so if we can solve that here with a `datasets` package, that'd be my preference.

rgommers commented:
> I think the dataset API should be abstracted away from files. Users should see data and python objects, not filenames.

Yes, I agree.

> I think if you wanted to get away from requests, you might have better effect making patches to pooch and suggesting them in PRs.

There's no need to specifically avoid `requests`. I saw @ilayn's comment higher up, but `urllib` may get deprecated and itself recommends `requests`. If we optionally depend on Pooch, this is just something we shouldn't care about (and if we did, `requests` is the gold standard).

ilayn commented:
> urllib may get deprecated and itself recommends requests

That's really good news for me. Miraculously I missed the whole PSF takeover. Makes many things so much simpler.

AnirudhDagar commented:
Thanks all for the feedback, completely understand why it makes sense to have pooch as an optional dependency. Does this mean that `scipy` should pass all the tests without pooch after build (we can somehow skip datasets module tests)? Or can we add pooch as a requirement for the tests on the CI and then naturally tests should be fine (like done in the PR at the moment)?

I guess a user won't care about tests and it is fine to hard depend on pooch for tests to pass, as we do depend in the case of `matplotlib` as well.

When using SciPy, the dependency still remains optional, until someone actually imports the datasets module or calls one of the datasets functions, all is good. IIUC when `scipy.datasets` is imported, an `importerror` should be raised if pooch is missing, and at the same time ask the user to install it.

rgommers commented:
> I guess a user won't care about tests and it is fine to hard depend on pooch for tests to pass, as we do depend in the case of `matplotlib` as well. (on a second thought, not sure about this)

It's fine to have a hard dependency on Pooch for the test suite. It'll be hard to avoid, too many tests will depend on it. Matplotlib is a different story - it's literally only a handful of tests that depend on it.

hmaarrfk commented:
Some build systems downstream might have restrictions on connecting to the internet. During their testing phase.

Finally, there can be spurious failures with downloading large datasets in CIs

We actually allow users to test without pooch install:

https://github.com/scikit-image/scikit-image/blob/725e8f6bcccc7d4e78ba0d64168f8ed76b6dc818/skimage/data/_fetchers.py#L143

hmaarrfk commented:
i mostly wanted to point to our implementation rather than to say you should do.

rgommers commented:
> Some build systems downstream might have restrictions on connecting to the internet. During their testing phase.

This is easy to avoid by downloading in the install phase in CI scripts (more efficient anyway probably). So when you do:
```
python dev.py --build-only   # or `pip install .`
```
to install SciPy, then right after do whatever the command will be to download all datasets used in CI.

hmaarrfk commented:
> to install SciPy, then right after do whatever the command will be to download all datasets used in CI. 

I believe some are even restricted against doing this.

I think one example is debian.

Anyway, what we did was to skip any test where pooch failed for whatever reason.

AnirudhDagar commented:
> Anyway, what we did was to skip any test where pooch failed for whatever reason.

This sounds like a good idea to me if we decide to skip only if the failure is due to a problem with pooch downloading the dataset.

rgommers commented:
> I think one example is debian.

It must be allowed to run multiple commands I'd think. We already require:
```
git clone ...
git submodule update --init
```
That should be the "obtain sources" step. If needed we could make it possible to run some script straight after without first building SciPy.

hmaarrfk commented:
Debian: They include their files in a mirror of your repo: https://salsa.debian.org/python-team/packages/scipy/-/blob/master/debian/rules
Fedora: Uses the url provided from github https://src.fedoraproject.org/rpms/scipy/blob/rawhide/f/scipy.spec#_30

Our implementation on skipping tests upon failure. https://github.com/scikit-image/scikit-image/blob/main/skimage/data/_fetchers.py#L217

The final case to consider, is that hammering raw.github is surely going to give you failures on builds due to github rejecting some of your requests. (3 platforms, 4+ python versions, staging & deploy = 24 builds)

to be honest, I think there are many ways to skip tests afterwards, and I know you are all very responsive to people raising build issues (and involved in the builds many times).

I wouldn't hold off because of the "nits" I am raising.

Users have explicitly asked that we make pooch an optional dependency.
We haven't heard too much from a testing side.

AnirudhDagar commented:
Thanks Ralf, yes I was able to reproduce this at a small isolated demo scale for pooch [here](https://github.com/AnirudhDagar/random-tests/runs/7219790510?check_suite_focus=true). It raises on the pooch import which as you suggested can be silenced, wrapping in a context manager filtering out the warning.

Do you think something like this is fine?

```python
import warnings

with warnings.catch_warnings():
    warnings.filterwarnings(action='ignore', category=DeprecationWarning)
    import pooch
```

rgommers commented:
Yes, that should do the job @AnirudhDagar.

If you can add that fix, and also have a look at if the remaining linting issues can be resolved without making the code worse, then CI should turn green.

AnirudhDagar commented:
For the lint errors:
```
scipy/datasets/__init__.py:33:1: E402 module level import not at top of file
scipy/datasets/_fetchers.py:85:80: E501 line too long (91 > 79 characters)
scipy/datasets/_registry.py:9:80: E501 line too long (85 > 79 characters)
scipy/datasets/_registry.py:10:80: E501 line too long (82 > 79 characters)
scipy/datasets/_registry.py:11:80: E501 line too long (82 > 79 characters)
scipy/datasets/_registry.py:15:80: E501 line too long (100 > 79 characters)
scipy/datasets/_registry.py:16:80: E501 line too long (91 > 79 characters)
scipy/datasets/_registry.py:17:80: E501 line too long (93 > 79 characters)
```

[__init__.py E402](https://github.com/AnirudhDagar/scipy/blob/b8098b4b9922d51e6aa7d691e86e4e36ea4a6775/scipy/datasets/__init__.py#L33) import is also something that is exactly like other modules in SciPy, so i feel it is okay to ignore.
[_fetchers.py E501](https://github.com/AnirudhDagar/scipy/blob/b8098b4b9922d51e6aa7d691e86e4e36ea4a6775/scipy/datasets/_fetchers.py#L85) is again a link in the doc which can be ignored.
[All registry E501s](https://github.com/AnirudhDagar/scipy/blob/scipy-datasets/scipy/datasets/_registry.py#L9-L17) are links which can be ignored for now.

WarrenWeckesser commented:
Sorry for not getting involved sooner.  It is great to see this moving forward.

However, I still believe we should have a small collection of datasets *included with SciPy*, that do not have to be downloaded on demand.  We should maintain a small set of datasets that will *just work* in the Examples sections of our docstrings without requiring an internet connection.  I think the current datasets that we have, plus something like a snapshot of the sunspot data, and a widely used multivariate dataset (not sure what that might be), would be easy to maintain and would not increase the source distribution by much.  It also cuts down on internet use for data loading when building the docs in CI.

This wouldn't necessarily change the API in the current PR, except that one does not get an error when accessing one of the pre-installed datasets when Pooch is not installed.

rgommers commented:
@WarrenWeckesser it's worth considering if any datasets need to still be included in the repo (`ascent` is maybe the best candidate), but "all we have now + some more" isn't very compelling imho. Have a look at `face` for example: it's 1.6 MB, and is only used in 2 docstring examples. Both sdist and wheel actually get smaller by that amount, because the file is already compressed. 

Excluding the three image files and building a wheel with `python -m build` shows the size on Linux goes from 23.5 MB to 21.6 MB, which is significant. For other data files:

```
$ find . -name "*.npz" | xargs du -sch
36K     ./linalg/tests/data/carex_19_data.npz
160K    ./linalg/tests/data/carex_18_data.npz
32K     ./linalg/tests/data/carex_20_data.npz
36K     ./linalg/tests/data/carex_15_data.npz
16K     ./linalg/tests/data/carex_6_data.npz
4.0K    ./linalg/tests/data/gendare_20170120_data.npz
164K    ./fftpack/tests/fftw_double_ref.npz
292K    ./fftpack/tests/fftw_longdouble_ref.npz
100K    ./fftpack/tests/fftw_single_ref.npz
12K     ./fftpack/tests/test.npz
24K     ./spatial/tests/data/degenerate_pointset.npz
576K    ./stats/_sobol_direction_numbers.npz
56K     ./special/tests/data/gsl.npz
204K    ./special/tests/data/local.npz
1.3M    ./special/tests/data/boost.npz
4.0K    ./interpolate/tests/data/bug-1310.npz
588K    ./sparse/linalg/tests/propack_test_data.npz
4.0K    ./sparse/tests/data/csc_py3.npz
4.0K    ./sparse/tests/data/csc_py2.npz
3.5M    total
```

```
$ find . -name "*.txt" | xargs du -sch
4.0K    ./spatial/qhull_src/COPYING.txt
4.0K    ./spatial/tests/data/pdist-correlation-ml.txt
4.0K    ./spatial/tests/data/pdist-hamming-ml.txt
12K     ./spatial/tests/data/random-int-data.txt
176K    ./spatial/tests/data/pdist-chebyshev-ml-iris.txt
4.0K    ./spatial/tests/data/pdist-spearman-ml.txt
8.0K    ./spatial/tests/data/cdist-X1.txt
8.0K    ./spatial/tests/data/random-bool-data.txt
52K     ./spatial/tests/data/pdist-double-inp.txt
176K    ./spatial/tests/data/pdist-jensenshannon-ml-iris.txt
16K     ./spatial/tests/data/iris.txt
180K    ./spatial/tests/data/pdist-minkowski-3.2-ml-iris.txt
176K    ./spatial/tests/data/pdist-correlation-ml-iris.txt
4.0K    ./spatial/tests/data/pdist-jensenshannon-ml.txt
12K     ./spatial/tests/data/random-uint-data.txt
180K    ./spatial/tests/data/pdist-minkowski-5.8-ml-iris.txt
180K    ./spatial/tests/data/pdist-euclidean-ml-iris.txt
176K    ./spatial/tests/data/pdist-cityblock-ml-iris.txt
4.0K    ./spatial/tests/data/pdist-euclidean-ml.txt
176K    ./spatial/tests/data/pdist-cosine-ml-iris.txt
52K     ./spatial/tests/data/pdist-boolean-inp.txt
4.0K    ./spatial/tests/data/pdist-chebyshev-ml.txt
4.0K    ./spatial/tests/data/pdist-jaccard-ml.txt
12K     ./spatial/tests/data/cdist-X2.txt
80K     ./spatial/tests/data/random-double-data.txt
4.0K    ./spatial/tests/data/pdist-cityblock-ml.txt
4.0K    ./spatial/tests/data/selfdual-4d-polytope.txt
180K    ./spatial/tests/data/pdist-seuclidean-ml-iris.txt
4.0K    ./spatial/tests/data/pdist-minkowski-3.2-ml.txt
4.0K    ./spatial/tests/data/pdist-cosine-ml.txt
4.0K    ./spatial/tests/data/pdist-seuclidean-ml.txt
4.0K    ./io/matlab/tests/data/japanese_utf8.txt
4.0K    ./ndimage/tests/data/label_strels.txt
8.0K    ./ndimage/tests/data/label_results.txt
4.0K    ./ndimage/tests/data/label_inputs.txt
1.9M    total
```

Other contributors:
- other `.dat` files aside from the 3 image files: 1.8M
- `.mat`: 0.46M total
- `.npy`: 0.41M total
- `.arff`: 0.26M total
- `.sav`: 0.29M total

> would not increase the source distribution by much

My goal would be to cut the wheels in size by as much as possible, since they're very much overweight, rather than "not increase much".

rgommers commented:
How about this: we merge this PR as is, and that gives us two releases to detect potential issues before the datasets in `misc` are removed. If it does give hiccups, we keep `ascent` and `ecg` as builtins in `scipy.datasets`, those have a reasonable size/utility tradeoff. And for the two `signal` examples that use `face`, we just switch those to `ascent`. 

> Eventually, it would be useful to have prepared a set of instructions for manually loading the cache of datasets, for our friends who work in remote or secure areas without continuous internet access.

Good point, this would be useful. I spent some time searching the scikit-learn issue tracker, they don't have an issue related to this (and only 40 issues related to datasets as a whole after ~7 years, which is a good sign). But it'd be a good thing to do anyway.

AnirudhDagar commented:
> Eventually, it would be useful to have prepared a set of instructions for manually loading the cache of datasets, for our friends who work in remote or secure areas without continuous internet access.

I can work on adding these doc instructions as a follow-up.

AnirudhDagar commented:
Thanks @rgommers!
> And those READMEs should probably point back to the SciPy docs page for the scipy.datasets submodule, which then explains the mechanics of how these datasets/repos are used, retrieved, cached, etc.

Good point about explaining the mechanics of how the datasets are fetched and cached using pooch. I have not mentioned that in this PR yet, so I'll update the docs there.

Regarding readme files of the dataset repos, should I add a future link? For example the following ascent doc link is obviously broken at the moment and will be applicable after the 1.10 release only. https://docs.scipy.org/doc/scipy/reference/generated/scipy.datasets.ascent.html#scipy.datasets.ascent

I'll send PRs on the three dataset repos.

rgommers commented:
> Regarding readme files of the dataset repos, should I add a future link?

Pointing to http://scipy.github.io/devdocs/ would be good

AnirudhDagar commented:
@rgommers I've rebased the PR, updated the dataset links and added documentation on usage of datasets. I've also sent PRs to the dataset repos adding a basic description.

AnirudhDagar commented:
Fixed the merge conflict due to #16739.

rgommers commented on scipy/datasets/__init__.py:
```diff
@@ -12,6 +12,47 @@
+for example `scipy.datasets.face` files live at
+https://github.com/scipy/dataset-face.
+The `scipy.datasets` submodule utilizes and depends on
+`Pooch <https://www.fatiando.org/pooch/latest/>`_, a python package built to
```
style: capitalize Python

rgommers commented on scipy/datasets/__init__.py:
```diff
@@ -12,6 +12,47 @@
+^^^^^^^^^^^^^^^^^
+
+SciPy dataset files are stored within individual github repositories under
+the SciPy Org, following a naming convention as ``'dataset-<name>'``,
```
style: "Org" -> "GitHub organization"

rgommers commented on scipy/datasets/__init__.py:
```diff
@@ -12,6 +12,47 @@
+
+A registry of all the datasets, essentially a mapping of filenames
+with their SHA256 hash and repo urls are maintained,
+which pooch uses to handle and verify the downloads
```
style: capitalize Pooch

rgommers commented on scipy/datasets/__init__.py:
```diff
@@ -12,6 +12,47 @@
+For Unix::
+
+    '~/.cache/scipy-data' or the value of the XDG_CACHE_HOME environment
+    variable, if defined.
```
I tested this, and imho it looks better here than for scikit-image:
```
$ ls ~/.cache/sci
scikit-image/ scipy-data/

$ ls ~/.cache/scipy-data/
ascent.dat  ecg.dat  face.dat

$ ls ~/.cache/scikit-image/
0.18.1
$ ls ~/.cache/scikit-image/0.18.1/data/
astronaut.png                      coffee.png                         lbpcascade_frontalface_opencv.xml  page.png
brick.png                          coins.png                          lfw_subset.npy                     phantom.png
camera.png                         color.png                          logo.png                           README.txt
cell.png                           grass.png                          microaneurysms.png                 retina.jpg
chelsea.png                        gravel.png                         moon.png                           rocket.jpg
chessboard_GRAY.png                horse.png                          motorcycle_disp.npz                text.png
chessboard_RGB.png                 hubble_deep_field.jpg              motorcycle_left.png                
clock_motion.png                   ihc.png                            motorcycle_right.png
```

This means that for scikit-image you get multiple copies on your machine, because it includes the version number. Data files should not be changing; if they do then any conflict can be resolved by either not reusing the same name or by clearing the cache (should work, unless there's a dramatic change). I don't think we have ever renamed a data file though.

rgommers commented on scipy/datasets/tests/test_data.py:
```diff
@@ -0,0 +1,48 @@
+    return pooch.file_hash(path) == expected_hash
+
+
+def test_download_all():
```
I noticed an issue here: the other tests are dependent on this one, and hence the tests will fail if run in parallel. E.g., try (after clearing the cache):
```
$ python dev.py test -s datasets --parallel 4
```
To address this, we may want to start using [pytest-dependency](https://pytest-dependency.readthedocs.io/en/stable/about.html). I think we perhaps may have other uses for those in the future (e.g. avoiding a ton of failures for highly parametrized tests).

Alternatively, all these tests can be put in a class with a `download_all()` call in the `setup` method (or whatever the thing is called that is executed once, rather than once per test.

rgommers replied:
Shouldn't we have a `scipy.datasets.download_all()` function by the way?

AnirudhDagar replied:
I see your point, but these functions are also supposed to download the datasets individually if there is no cached dataset present. Hence I don't think there is any dependency.

In fact, now that i think more about this, we should have two cases:
1. One where we actually use the cache for testing the datasets. In that case, indeed the dependency will be on `test_download_all` and pooch should utilize the cache.
2. Second is where we manually clear cache and then run tests for each dataset. In that case, each dataset will use pooch to download the files.

This way, we'll be testing all possible cases for dataset file retrieval.

`python dev.py test -s datasets --parallel 4` works just fine for me, but I'm curious if this can cause some sort of corruption (since more than one function may be downloading the file and writing/overwriting the contents at the same time. I'm guessing here).

> scipy.datasets.download_all()

Yeah, honestly, I see the use of that only once in the lifetime of scipy installation for a user, that too when someone wants to make sure that their SciPy installation always works without the internet connection. Seems like a very special case, but it wouldn't really hurt to add this utility function. Can be discussed, should be an easy addition with no maintenance anyway.

rgommers replied:
We discussed the race condition in person just now, and also that it's desirable not to download the same dataset twice. I'll wait for your next update here.

AnirudhDagar replied:
Done, except the second case is not being tested right now, will add that later when we have a utility function to clear all cache.

rgommers commented on scipy/datasets/__init__.py:
```diff
@@ -0,0 +1,82 @@
+   electrocardiogram - Load an example of a one-dimensional signal.
+
+
+Usage of Datasets
```
This isn't quite usage, more like a "how does it work under the hood". How about an actual `Usage` section above, which only explains that you call `dataset_name()` and it returns you the ndarray, dict or other Python data structure, and that it is cached so downloads over the network once? And then below that this `How dataset retrieval and storage works` section.

AnirudhDagar replied:
I've tried to improve this, how does it look now?

rgommers replied:
Looks better now. I pushed one commit with small tweaks like using the correct reST header underlines and removing the question mark from the last section title.

rgommers reviewed (commented):
I re-read all the comments and retested on Linux and macOS (shouldn't matter, since this is all pure Python) as well as verified the docs on how caching works. It all looks about ready to go to me. The test dependency comment for parallel tests is the most important one; that could fail CI straight away, so let's address it now.

rgommers commented:
Follow-ups we discussed:

- [ ] add `download_all`
- [ ] add something like a `clear_cache` function, so the user doesn't have to manually hunt for a platform-specific path
- [ ] revisit the comment higher up about the needs of a packager needing to adhere to Debian rules: we need some separate script or package to download everything at once before scipy is built
- [ ] skipping tests if Pooch is not available or there is no network connection

These can all be done after merging this PR.

AnirudhDagar commented on scipy/datasets/tests/test_data.py:
```diff
@@ -19,37 +22,52 @@ def _has_hash(path, expected_hash):
+
+        yield
 
-    assert len(os.listdir(data_dir)) >= len(registry)
```
@rgommers As this was converted to a fixture, I had to move the assert out into a new `test_existence_all` because it is not a good practice to add asserts during test_setup or test_teardown phase inside a fixture.

AnirudhDagar commented:
I've addressed the review comments (major changes to test_data.py design); I'll also open a new issue to track all the follow-up tasks related to SciPy datasets.

AnirudhDagar commented:
Btw before this is merged (no dependency though), just a small reminder about the PRs on the dataset repos. LMK if they are fine or need any improvement:
* scipy/dataset-ecg#1
* scipy/dataset-ascent#1
* scipy/dataset-face#1

rgommers commented on scipy/datasets/tests/test_data.py:
```diff
@@ -0,0 +1,73 @@
+
+        # test_teardown phase
+
+        # TODO: Clear all cache from test_download_all
```
For the follow-up PR on this: this isn't exactly right, the test should not clear the cache, otherwise we're re-downloading all data each time someone runs the test suite locally. You can give it an `xslow` marker to do that, so it's only run very occasionally.

AnirudhDagar replied:
That's a good idea and would be better to implement that marking the test as slow.

rgommers reviewed (approved):
Okay, I think this is in very good shape and the follow-ups are clear. Thanks for all the hard work @AnirudhDagar and @WarrenWeckesser, and everyone else for reviews and feedback!

@AnirudhDagar can you add a release note entry for this in https://github.com/scipy/scipy/wiki/*Release-Note-Entries-for-SciPy-1.10.0?

AnirudhDagar commented:
Thanks again @WarrenWeckesser for the initial work, and thanks @rgommers, for helping me through all of this.
I'll update the release notes and get started with the follow-ups. :)

oscarbenjamin commented:
This PR seems to be the cause of a CI failure in SymPy that is somehow related to importing polygamma https://github.com/sympy/sympy/pull/23932. Not sure yet why that would be affected but @bjodah bisected to here.

oscarbenjamin commented:
Maybe the problem is that this does not install the pooch dependency:
```
pip install git+https://github.com/scipy/scipy@main
```
Maybe the dependency should be added in `setup.py`.

AnirudhDagar commented:
Thanks @oscarbenjamin for the report. Yes, `pooch` is an optional dependency, and thus we don't require it. I'm curious, does SymPy use the `datasets` module already? Or are you running all the SciPy tests in SymPy CI (I've not checked SymPy CI yet)? This warning should only hit if you try to import `scipy.datasets` or use a method from that module.

All other sections/submodules within SciPy can still be used without installing/requiring `Pooch`.

AnirudhDagar commented:
Ohh interesting, is it just the install from git that is triggering the issue?

oscarbenjamin commented:
No SymPy does not use the datasets module. Somehow this PR has affected SymPy importing other things from SciPy as part of lambdify. The code that does this is quite convoluted so it's not easy to say exactly what import statement is being used:
https://github.com/sympy/sympy/blob/master/sympy/utilities/lambdify.py

oscarbenjamin commented:
> Ohh interesting, is it just the install from git that is triggering the issue?

Yes, somehow SymPy CI fails with SciPy main since this PR was merged. More discussion in https://github.com/sympy/sympy/pull/23932.

AnirudhDagar commented:
The file indeed is convoluted, and I need to look closely, but from the first glance, maybe it is L106 importing everything from SciPy. 

https://github.com/sympy/sympy/blob/b1f35f8daf5efa9f801f905cf1037000d69b0278/sympy/utilities/lambdify.py#L106

I'll be happy to help out and dig deeper in SymPy code later in the day (slightly busy atm); I hope it's not too urgent, and is fine if I reply late evening or by tomorrow morning. Thanks!

rgommers commented:
Yes, `from scipy import *` will import `scipy.datasets, because `datasets` is added to `__all__`. I suggest fixing it by removing `from scipy import *`, that is _really_ bad practice and it's then expected to pull in optional components. There is also nothing much of interest in the main `scipy` namespace that even makes it useful to star-import.

oscarbenjamin commented:
To be clear since I haven't managed to build scipy main locally does this PR make it an importerror to do `from scipy import *` without having pooch installed?

AnirudhDagar commented:
> To be clear since I haven't managed to build scipy main locally does this PR make it an importerror to do from scipy import * without having pooch installed?

Importing everything will also include `scipy.datasets` submodule, which depends on `pooch` and we don't make it a requirement so yes, importing everything will prompt an `import error` and ask you to install pooch. But as Ralf said it's a bad practice to `import *` and that's not something you want to do.

oscarbenjamin commented:
I agree that the sympy code should be changed but does this not sound like a scipy regression to you?

rgommers commented:
Not really, no. There's many non-idiomatic things one can do, and in this case:

- we have documented importing guidelines at http://scipy.github.io/devdocs/reference/index.html#guidelines-for-importing-functions-from-scipy,
- not adding `datasets` to `__all__` also seems wrong - it would avoid your SymPy issue, but only by completely omitting the optional submodule from the list of public objects. And if we'd try-except `import pooch` in `scipy/__init__.py` I'm sure other people will be unhappy by the extra overhead/import that they didn't ask for.

oscarbenjamin commented:
The sympy code should be changed in any case (although I would prefer not to have to do that at release candidate stage) so let's not be concerned with that here.

I imagine that `from scipy import *` is something that has worked just fine for a long time even if you don't consider it idiomatic and I expect that a lot of code depends on it. Obviously it's your call but to me this seems like a fairly gratuitous breakage.

rgommers commented:
Okay, let's open an issue for it, tag it with the `1.10.0` milestone, and see if we need to do something here before the next release. I don't think we want to change it right now in `main`, that is actually the right time to make changes and smoke out other possible issues.

AnirudhDagar commented:
It honestly will make a lot of users unhappy, not sure if there is a way to delay the import error raised.

Something like this inside `datasets.__init__.py` would avoid the error, and we can shift the error to `_fetchers.py` but it is obv not the best idea because now the user who wants to use this will need to install and reimport:

```
try:
    import pooch
    from ._fetchers import face, ascent, electrocardiogram  # noqa: E402
    __all__ = ['ascent', 'electrocardiogram', 'face']

except:
    warnings.warn("Pooch not found! Please install pooch before using scipy.datasets."
                  "After installing pooch, try importing scipy.datasets again .....")

from scipy._lib._testutils import PytestTester
test = PytestTester(__name__)
del PytestTester
....
```

rgommers commented:
> `After installing pooch, try importing scipy.datasets again`

A second import statement for the same module is a no-op, so that may not do what you intend.

AnirudhDagar commented:
Ohhh I see, that's correct, will not work until the user restarts the process. I don't have any ideas atm to avoid this.

oscarbenjamin commented:
> not sure if there is a way to delay the import error raised.

There are lots of ways that you can make this work if you want to make it work. The code here goes out of its way to raise ImportError. If you wanted the module to be importable but without some of its features when pooch is not installed then you just do something like:
```python
try:
    import pooch
except ImportError:
    pooch = None
    data = None
else:
    data = pooch.create(...)

def fetch_data(...):
    if data is None:
        raise an error saying that pooch is not installed
    return data.fetch(...)
```
Then everywhere you currently call `data.fetch` you can just call `fetch_data` instead.

> will not work until the user restarts the process. I don't have any ideas atm to avoid this

In general it is always necessary to restart the process after install or uninstalling things so that doesn't seem like a big deal. If you did want to avoid that though then you would just move `import pooch` into the `fetch_data` function rather than being at top-level.

AnirudhDagar commented:
Thanks @oscarbenjamin this is good and should work. Sorry for the negligence on my side, this was quite straightfwd. Moving pooch import into `scipy.datasets._fetchers.py` will do the trick. @rgommers do you think it is okay to not raise a warning/error if someone does `from scipy import datasets`? Now that I think it of it, i feel it should have been that way from the start. Got that wrong :/

If that's okay, I'll send a PR to fix this now.

asmeurer commented:
I disagree that SymPy should change here. `from mod import *` is how `lambdify` works, to import all the names into the namespace (see the "how it works" section here https://docs.sympy.org/dev/modules/utilities/lambdify.html#sympy.utilities.lambdify.lambdify).

If SciPy is hard importing an external library with `from scipy import *`, then that library has effectively become a hard dependency of scipy. SymPy is hardly going to be the only instance doing an `import *` so this will likely affect more than just it.

I'm actually surprised that you can add something to `__all__` without adding it to the main `import scipy`. I never realized `__all__` worked like that.

rgommers commented:
> I disagree that SymPy should change here. `from mod import *` is how `lambdify` works, to import all the names into the namespace (see the "how it works" section here https://docs.sympy.org/dev/modules/utilities/lambdify.html#sympy.utilities.lambdify.lambdify).

I can't understand from the link why `from scipy import *` is needed. You're not also recursively star-importing `scipy.submodule`, right? 99.x% of scipy's functionality is present in those submodules, there's only a handful of items in the main namespace. So it looks to me like this is not actually necessary. And it's most definitely bad practice.

asmeurer commented:
It's pulling everything from scipy and scipy.special. I guess as Oscar showed at https://github.com/sympy/sympy/pull/23945, we really only need scipy.special. 

> And it's most definitely bad practice.

`import *` is bad practice is certain situations, but perfectly acceptable in others. If you really believe that it's universally bad, then why does scipy even define `__all__`?

rgommers commented:
> then why does scipy even define `__all__`?

`__all__` has always been used to define all public symbols, and we use it also to verify that all public functions are present in the API docs.
